### PR TITLE
Run E2E tests in Windows nodes

### LINKED ISSUE #277: Run E2E tests in Windows
After enabling the multidistro build we have to run the E2E tests in Windows nodes, the plan is:

- Enable OpenSSH in the Windows nodes, luckily the GKE Windows team made a lot of effort setting it up in a GCE Kubernetes cluster that has Windows nodes, we don't have to do anything here. Unfortunately this isn't set up in a GKE cluster by default so we'll need to test in a GCE cluster.
- Enable HyperV, it's a Windows feature that enables the creation of Virtual Hard Disks (VHD), with this new API we can create virtual volumes that will be attached to the LVP daemonset and tested in the E2E tests.
- Updating the E2E tests to execute Powershell scripts through SSH, the scripts will create the VHDs and any setup that needs to be done in the Windows node.

### PR DESCRIPTION
**Pull request by mauriciopoppe**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
/kind feature

**What this PR does / why we need it**:
Updates the E2E test suite to run the tests in Windows nodes, steps:

- Implementation of a new host executor for Windows nodes that connects through `gcloud compute ssh` to the Windows nodes to execute powershell commands
- Implementation of a VHD script generator, VHDs are created and mounted to the discovery directory in the host, VHDs are created with the host executor and it comes as a drop-in replacement for the Linux host executor when the test runs with Windows nodes

**Which issue(s) this PR fixes**:

Fixes #277

**Release note**:
```
NONE
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,6 +5,7 @@ go 1.16
 require (
 	github.com/ghodss/yaml v1.0.0
 	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
+	github.com/kubernetes-csi/csi-proxy/client v1.0.1
```
Comment (jingxu97) [inline go.mod:8]: 1.0.2?

File: hack/e2e.sh [hunk 2]
```diff
@@ -316,9 +326,8 @@ fi
...
 go run $ROOT/hack/e2e.go -- "${kubetest_args[@]}" \
-    --deployment "$DEPLOYMENT" \
+    --deployment="$DEPLOYMENT" \
     --up \
     --down \
-    --test-cmd bash \
-    --test-cmd-args="$ROOT/hack/run-e2e.sh" \
-    "$@"
+    --test-cmd="$ROOT/hack/run-e2e.sh" \
+    "${@:-}"
```
Comment (jingxu97) [inline hack/e2e.sh:333]: what the difference between this and before?

File: hack/run-e2e.sh [hunk 3]
```diff
@@ -80,10 +82,21 @@ fi
...
+    REGISTRY=${REGISTRY:-gcr.io/${PROJECT}}
+    IMAGE=local-volume-provisioner
+
+    REGISTRY=${REGISTRY} \
+        VERSION=$VERSION \
+        CONFIRM=false \
+        ALLOW_UNSTABLE=true \
+        ALLOW_DIRTY=true \
+        LINUX_ARCH="amd64" \
+        WINDOWS_DISTROS="ltsc2019" \
```
Comment (jingxu97) [inline hack/run-e2e.sh:96]: so we choose ltsc2019 to test, is this configurable?

File: go.mod (hunk 1 above)
Comment (mauriciopoppe) [reply go.mod:8]: should be fixed after https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/pull/276/ is merged

File: hack/run-e2e.sh (hunk 3 above)
Comment (mauriciopoppe) [reply hack/run-e2e.sh:96]: it is configurable but it the test-infra job has to be changed too in, PR for test infra https://github.com/kubernetes/test-infra/pull/24241/files

File: hack/e2e.sh (hunk 2 above)
Comment (mauriciopoppe) [reply hack/e2e.sh:333]: before `bash` was the test command, now it's run-e2e.sh and instead more args come to it from https://github.com/kubernetes/test-infra/pull/24241

**Comment by mauriciopoppe** [comment]
/test pull-sig-storage-local-static-provisioner-e2e-windows-2019

Windows nodes are not coming up, debugging logs in https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/880

File: hack/e2e.sh [hunk 4]
```diff
@@ -131,6 +132,9 @@ while getopts "h?" opt; do
     esac
 done
 
+# TODO(mauriciopoppe): remove this line after retry
```
Comment (jingxu97) [inline hack/e2e.sh:135]: what is retry?

File: hack/release.sh [hunk 5]
```diff
@@ -91,7 +91,7 @@ ALLOW_OVERRIDE=${ALLOW_OVERRIDE:-}
 SKIP_BUILD=${SKIP_BUILD:-}
 SKIP_PUSH_LATEST=${SKIP_PUSH_LATEST:-}
 LINUX_ARCH=${LINUX_ARCH:-amd64 arm arm64 ppc64le s390x}
-WINDOWS_DISTROS=${WINDOWS_DISTROS:-ltsc2019 1909 2004 20H2}
+WINDOWS_DISTROS=${WINDOWS_DISTROS:-ltsc2019 1909 2004 20H2 ltsc2022}
```
Comment (jingxu97) [inline hack/release.sh:94]: I think SAC 1909 and 2004 is end of service.

File: test/e2e/dev/main.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	klog.InitFlags(nil)
+	defer klog.Flush()
+
+	flag.Parse()
+
+	klog.Infof("Remotely executing a program!")
+
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "e2e-test-mauriciopoppe-windows-node-group-l64t",
```
Comment (jingxu97) [inline test/e2e/dev/main.go:40]: update the name?

File: test/e2e/dev/main.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		},
+	}
+	hostExec := windows.NewHostExec()
+	result, err := hostExec.Execute("echo $env:USERPROFILE", node)
+	if err != nil {
+		panic(err)
+	}
+	e2estorageutils.LogResult(result)
+
+	vhd := windows.NewVHD("C:\\var\\lib\\kubelet\\plugins\\testplugin-0.csi.io\\disk-dev-9.vhdx", 1024*1024*1024)
```
Comment (jingxu97) [inline test/e2e/dev/main.go:50]: why use disk-dev-9?

File: test/e2e/dev/main.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	if err != nil {
+		panic(err)
+	}
+	result, err = hostExec.Execute(vhdStage, node)
+	if err != nil {
+		panic(err)
+	}
+	e2estorageutils.LogResult(result)
+
+	vhdPublish, err := vhd.PublishScript("C:\\var\\lib\\kubelet\\plugins\\testplugin-0.csi.io\\mount-9")
```
Comment (jingxu97) [inline test/e2e/dev/main.go:62]: mount-9?

File: test/e2e/windows/vhd.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+// NewVHD creates a new VHD
+func NewVHD(path string, sizeBytes int) *VHD {
+	return &VHD{
+		Path:      path,
+		SizeBytes: sizeBytes,
+	}
+}
+
+// StageScript generates a script to initialize a VHD
+func (vhd *VHD) StageScript() (string, error) {
```
Comment (jingxu97) [inline test/e2e/windows/vhd.go:42]: what is the return string?

File: hack/e2e.sh (hunk 4 above)
Comment (mauriciopoppe) [reply hack/e2e.sh:135]: I’m going to remove this line, and lease ignore it

File: hack/release.sh (hunk 5 above)
Comment (mauriciopoppe) [reply hack/release.sh:94]: great, I’m going to remove them

File: test/e2e/dev/main.go (hunk 6 above)
Comment (mauriciopoppe) [reply test/e2e/dev/main.go:40]: thanks, fixed

File: test/e2e/dev/main.go (hunk 7 above)
Comment (mauriciopoppe) [reply test/e2e/dev/main.go:50]: this is just a dev script, it’s not used anywhere

File: test/e2e/dev/main.go (hunk 8 above)
Comment (mauriciopoppe) [reply test/e2e/dev/main.go:62]: same as above

File: test/e2e/windows/vhd.go (hunk 9 above)
Comment (mauriciopoppe) [reply test/e2e/windows/vhd.go:42]: this file generates scripts to be used by a windows executor, the executor works with strings

File: test/e2e/dev/main.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	flag.Parse()
+
+	klog.Infof("Remotely executing a program!")
+
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "e2e-test-mauriciopoppe-windows-node-group-l64t",
+		},
+	}
+	hostExec := windows.NewHostExec()
```
Comment (jingxu97) [inline test/e2e/dev/main.go:43]: this program is only for windows case, should it in windows subdir?

File: test/e2e/e2e_test.go [hunk 11]
```diff
@@ -356,7 +399,8 @@ var _ = utils.SIGDescribe("PersistentVolumes-local ", func() {
 				for i := 0; i < numConcurrentPods; i++ {
 					pvcs := []*v1.PersistentVolumeClaim{}
 					for j := 0; j < volsPerPod; j++ {
-						pvc := e2epv.MakePersistentVolumeClaim(makeLocalPVCConfig(config, DirectoryLocalVolumeType), config.ns)
+						pvcConfig := makeLocalPVCConfig(config, DirectoryLocalVolumeType)
```
Comment (jingxu97) [inline test/e2e/e2e_test.go:402]: what's the difference with before?

File: test/e2e/e2e_test.go [hunk 12]
```diff
@@ -392,7 +436,8 @@ var _ = utils.SIGDescribe("PersistentVolumes-local ", func() {
 			}()
 
 			By("Waiting for all pods to complete successfully")
-			err := wait.PollImmediate(time.Second, 5*time.Minute, func() (done bool, err error) {
+			timeout := 10 * time.Minute
```
Comment (jingxu97) [inline test/e2e/e2e_test.go:439]: timeout is changed due to windows test?
we could make it different for windows and linux case.

File: test/e2e/dev/main.go (hunk 10 above)
Comment (mauriciopoppe) [reply test/e2e/dev/main.go:43]: moved to `test/cmd/remote-exec-dev/`, `test/e2e/windows` is a library that has to be used from a main package

File: test/e2e/e2e_test.go (hunk 12 above)
Comment (mauriciopoppe) [reply test/e2e/e2e_test.go:439]: sure, I changed it to be 10m for windows only

File: test/e2e/e2e_test.go (hunk 11 above)
Comment (mauriciopoppe) [reply test/e2e/e2e_test.go]: no difference, probably I did something in between like logging and forgot to make it like it was before

**Comment by mauriciopoppe** [comment]
/remove-lifecycle stale

I need to pick some commits from this PR and make the e2e tests work with a new impl of LVP making it run on Windows with HostProcess containers.

**Pull request closed without merging**